frontend/javascripts/oxalis/geometries/materials/plane_material_factory.js

```diff
@@ -51,11 +51,13 @@ export type Uniforms = {
 const DEFAULT_COLOR = new THREE.Vector3([255, 255, 255]);
 
 function sanitizeName(name: ?string): string {
-  // Make sure name starts with a letter and contains no "-" characters
   if (name == null) {
     return "";
   }
-  return `layer_${name.replace(/-/g, "_")}`;
+  // Variables must start with a-z,A-Z or _. Names can contain a-z,A-Z,0-9 or _.
+  // User variable names cannot start with gl_ or contain a double _.
+  // Base64 encode the layer name and remove = characters to make sure variable names are valid
+  return `layer_${btoa(name).replace(/=+/g, "")}`;
```

Comment:
Hmm, we base64 encoding loses the readability of the actual layer name (useful when debugging the generated shader code). Couldn't we adapt the original regex expression so that all disallowed characters are removed (or replaced)? E.g. `name.replace(/[^a-zA-Z0-9_]/g, "_")`. Additionally, we could always prepend a `_` character to guard against disallowed variable beginnings (e.g, `gl_` or numbers). Afterwards, `__` could be replaced with `_` and then we should be good?

Reply:
>  Couldn't we adapt the original regex expression so that all disallowed characters are removed (or replaced)?

I thought about this as well, but it doesn't work. Essentially, layer names will no longer be unique after the regex transformation. Layers `seg+type`, `seg-type` and `seg.+type` will all be converted to `layer_seg_type` and we will get name collisions.

We could prepend a random part computed deterministically from the layer name, but I'm not really convinced by this. Variable names would become even longer :/

>  readability of the actual layer name (useful when debugging the generated shader code)

My thinking was, that if it was really important what the actual layer name was, the debugger could simply run `btoa(orig_layer_name)` or `atob(base64string)` in the console.

Reply:
Ok, I see. I think, the chances of having seg+type, seg-type and seg.+type in the same dataset are quite low, but I agree that it's better to be on the safe side here :+1:

Reply:
Just an idea: What about URL encoding/escaping for this? e.g. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent

Reply:
From the mdn page:

> Not Escaped:
    A-Z a-z 0-9 **- _ . ! ~ * ' ( )**

so we would still need to treat these special chars separately.